wai-extra/Network/Wai/Parse.hs

```diff
@@ -351,23 +567,32 @@ data WTB = WTBWorking (S.ByteString -> S.ByteString)
          | WTBDone Bool
 wrapTillBound :: S.ByteString -- ^ bound
               -> Source
-              -> IO (IO S.ByteString, IO Bool) -- ^ Bool indicates if the bound was found
-wrapTillBound bound src = do
+              -> Maybe Int64
+              -> IO (IO S.ByteString, IO (Bool, Int64)) -- ^ Bool indicates if the bound was found
+wrapTillBound bound src max' = do
     ref <- newIORef $ WTBWorking id
-    return (go ref, final ref)
+    sref <- newIORef (0 :: Int64)
+    return (go ref sref, final ref sref)
   where
-    final ref = do
+    final ref sref = do
         x <- readIORef ref
         case x of
             WTBWorking _ -> error "wrapTillBound did not finish"
-            WTBDone y -> return y
+            WTBDone y -> do
+                siz <- readIORef sref
+                return (y, siz)
 
-    go ref = do
+    go ref sref = do
         state <- readIORef ref
         case state of
             WTBDone _ -> return S.empty
             WTBWorking front -> do
                 bs <- readSource src
+                cur <- atomicModifyIORef sref $ \ cur ->
```

Comment:
Probably want to either use a bang pattern on `cur` or switch to `atomicModifyIORef'` to avoid a memory leak in the `Nothing` case of `max'`.